src/rt/runtime.rs

```diff
@@ -99,21 +97,57 @@ impl Runtime {
     /// Runs the runtime on the current thread.
     pub fn run(&self) {
         scope(|s| {
-            for m in &self.machines {
-                s.builder()
-                    .name("async-std/machine".to_string())
-                    .spawn(move |_| {
-                        abort_on_panic(|| {
-                            let _ = MACHINE.with(|machine| machine.set(m.clone()));
-                            m.run(self);
+            let mut idle = 0;
+            let mut delay = 0;
+
+            loop {
+                // Get a list of new machines to start, if any need to be started.
+                for m in self.make_machines() {
+                    idle = 0;
+
+                    s.builder()
+                        .name("async-std/machine".to_string())
+                        .spawn(move |_| {
+                            abort_on_panic(|| {
+                                let _ = MACHINE.with(|machine| machine.set(m.clone()));
+                                m.run(self);
+                            })
                         })
-                    })
-                    .expect("cannot start a machine thread");
+                        .expect("cannot start a machine thread");
+                }
+
+                // Sleep for a bit longer if the scheduler state hasn't changed in a while.
+                if idle > 10 {
+                    delay = (delay * 2).min(10_000);
```

Comment:
Is perhaps crossbeam utils BackOff a better fit here than rolling it by hand?

Reply:
I figured if it was good enough for @stjepang it'll be good enough for me for now ;)